### PR TITLE
Bug fix: user avatar does not update on challenge completion

### PR DESCRIPTION
**Pull request by CHOOSEIT**
Fixes: #342 

This PR aims to fix the user avatar not updating properly on challenge completion.

image

As you can see with the line changes, this was not a straight-forward bug fix. 
We had not any updates on the user avatars. The user avatars were also not linked to any provider, hence, on challenge completion, we had no way to update every user avatar at once.

This PR creates a new provider `userCentauriPointsViewModelProvider`, that gets the centauri points of a user given its user id. This provider is refreshable, allowing us to watch on it in every user avatar.

Note that the user avatar had not access to the user id of the displayed user, so a considerable part of this PR is giving this access to this information.

### CODE REVIEW

File: test/models/ui/comment_details_test.dart
Comment (yoannLafore) [inline test/models/ui/comment_details_test.dart:56]: Very small suggestion, I found out that we could test inequality with the matchers `isNot(equals(***))`.
This would gives:
```suggestion
      expect(commentPost, isNot(equals(commentPostOther)));
```

I personnaly feels like it improves readability a little bit.
But this is just my point of view and I'd totally understand if you choose to stick with your approach :)

File: test/models/ui/ranking/ranking_element_details_test.dart
Comment (yoannLafore) [inline test/models/ui/ranking/ranking_element_details_test.dart:128]: If you choose to go with my previous suggestion, I think it could also be applied here.

```suggestion
      expect(rankingElementDetail1, isNot(equals(rankingElementDetail2)));
```

File: test/viewmodels/user_centauri_points_view_model_unit_test.dart
Comment (yoannLafore) [inline test/viewmodels/user_centauri_points_view_model_unit_test.dart:37]: I think `userRepo` isn't use in the tests. 
Thus I believe we can remove it.

```suggestion
    late FakeFirebaseFirestore fakeFireStore;

    late ProviderContainer container;

    late List<UserFirestore> availableUsers;

    setUp(() async {
      fakeFireStore = FakeFirebaseFirestore();

      availableUsers = FirestoreUserGenerator.generateUserFirestore(10);
      await setUsersFirestore(fakeFireStore, availableUsers);

      //Prepare the list of providers to override.
      container = ProviderContainer(
        overrides: [
          firestoreProvider.overrideWithValue(fakeFireStore),
        ],
      );
    });
```

**Review by yoannLafore — COMMENTED** [review]
Thank you for fixing this bug !
This effectively contributes to making our application more responsive to the user and so improves the overall UX.

Your code is of great quality with abondant documentation :+1: 
The PR overall looks good to me and I saw that you added some tests for the new view model which is great.

I have just made a few suggestion below.

Thanks for your great work :tada:

File: lib/utils/user_avatar_color.dart
Comment (gruvw) [reply lib/utils/user_avatar_color.dart:7]: btw in dart I think we should avoid class with only static members, check out https://dart.dev/tools/linter-rules/avoid_classes_with_only_static_members

File: lib/models/ui/user_avatar_details.dart
Comment (CHOOSEIT) [reply lib/models/ui/user_avatar_details.dart]: You are completely right ! 

Added in: [7a9eeaf](https://github.com/ProximaEPFL/proxima/pull/348/commits/7a9eeafec0ee9dfe285594e4d73d03bdfd407de1)

Thank you !

File: lib/utils/user_avatar_color.dart
Comment (CHOOSEIT) [reply lib/utils/user_avatar_color.dart]: You are completely right 💯 

I removed the class in : [4d15cbb](https://github.com/ProximaEPFL/proxima/pull/348/commits/4d15cbbb38197d4aae52aeb6720e5c40d40f3835)

Thank you !

File: lib/viewmodels/dynamic_user_avatar_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/dynamic_user_avatar_view_model.dart]: Yes, that's clearly more readable 😄 

Changed also in: [7a9eeaf](https://github.com/ProximaEPFL/proxima/pull/348/commits/7a9eeafec0ee9dfe285594e4d73d03bdfd407de1)

Thank you !

Comment (CHOOSEIT) [reply lib/viewmodels/dynamic_user_avatar_view_model.dart]: Yes, that's more concise 😄 

Changed in: [7a9eeaf](https://github.com/ProximaEPFL/proxima/pull/348/commits/7a9eeafec0ee9dfe285594e4d73d03bdfd407de1)

Thank you !

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart]: You are right ! It can create some confusion when using it.

Changed in: [f5fc5a7](https://github.com/ProximaEPFL/proxima/pull/348/commits/f5fc5a75ada6f66e41447466e0397ce4fa6c98f5)

Thank you !

File: test/models/ui/ranking/ranking_element_details_test.dart
Comment (CHOOSEIT) [reply test/models/ui/ranking/ranking_element_details_test.dart:128]: You are completely right.

Changed in [cefc851](https://github.com/ProximaEPFL/proxima/pull/348/commits/cefc85186f44b65ca6bdfe9f097aaea631128868)

I also added this matcher to the other tests in this file :)

Thank you !

File: test/models/ui/comment_details_test.dart
Comment (CHOOSEIT) [reply test/models/ui/comment_details_test.dart]: You are completely right.

Changed in [cefc851](https://github.com/ProximaEPFL/proxima/pull/348/commits/cefc85186f44b65ca6bdfe9f097aaea631128868)

I also added this matcher to the other tests in this file :)

Thank you !

File: test/models/ui/ranking/ranking_element_details_test.dart
Comment (CHOOSEIT) [reply test/models/ui/ranking/ranking_element_details_test.dart]: You are right !

Applied in: [c7be8de](https://github.com/ProximaEPFL/proxima/pull/348/commits/c7be8de3f35c0cf986dac822e0e5fe83c1b4c4b1)

Note that I also applied this change for the other tests in the file :)

Thank you !

File: test/viewmodels/user_centauri_points_view_model_unit_test.dart
Comment (CHOOSEIT) [reply test/viewmodels/user_centauri_points_view_model_unit_test.dart]: You are completely right !

Removed in [7126752](https://github.com/ProximaEPFL/proxima/pull/348/commits/7126752cd987a85b2aa4aeac237f6501bbe51336)

Thank you !

File: lib/utils/user_avatar_color.dart
Comment (yoannLafore) [reply lib/utils/user_avatar_color.dart]: I didn't know that either, thanks for pointing it out !

Also, thanks for making the change :+1:

File: test/models/ui/comment_details_test.dart
Comment (yoannLafore) [reply test/models/ui/comment_details_test.dart]: Perfect :ok_hand: 
Thank you !

**Review by yoannLafore — APPROVED** [review]
Thank you for bringing the modifications !
The PR looks good to me now :+1: 

Amazing job :tada:

**Comment by CHOOSEIT** [comment]
Thank you, @yoannLafore, for your review 🎉 

As this PR changes a lot of things, we are going to wait for @JoachimFavre's review before merging 💯

File: lib/viewmodels/users_ranking_view_model.dart
Comment (JoachimFavre) [inline lib/viewmodels/users_ranking_view_model.dart:87]: The following allows to slightly decrease code repetition:
```
    // Refresh the centauri points of the top users and the current user
    for (final user in topUsersFromDb + [currentUser]) {
      ref
          .read(
            userCentauriPointsViewModelProvider(user.uid).notifier,
          )
          .refreshWithCentauriPointsNumber(
            user.data.centauriPoints,
          );
    }
```

This is not a GitHub suggestion since I think it requires removing unused variables.

File: lib/utils/user_avatar_color.dart
Comment (JoachimFavre) [inline lib/utils/user_avatar_color.dart:5]: By humble curiosity, why did you do this refactor?

I'm not a big fan of its current form, because I feel like it adds some namespace pollution. If you don't want to put those lines back to the dynamic user avatar view-model, I think you should at least encapsulate everything inside a class `UserAvatarColor` with a private constructor.

File: lib/views/components/content/user_avatar/user_avatar.dart
Comment (JoachimFavre) [inline lib/views/components/content/user_avatar/user_avatar.dart:77]: You can make this code much simpler by taking advantage of `AsyncValue::value` which returns the previous value if it exists or null otherwise, and re-throws the error if the `AsyncValue` is in an error state; and with the fact that `centauriToUserAvatarColor` returns the loading value when the number of points given in argument is `null`.
```
    final backgroundColor = centauriToUserAvatarColor(
      userCentauriAsync.value,
      Theme.of(context).brightness,
    );

    return CircleAvatar(
      radius: radius,
      backgroundColor: backgroundColor,
      child: content,
    );
```

In fact, using this method, you can go back to the way this file originally was by adding very few lines.

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (JoachimFavre) [inline lib/viewmodels/user_centauri_points_view_model.dart:30]: If you decide to apply my previous comment, I'd recommend to also use the following here:
```suggestion
    state = const AsyncValue.loading();
    state = await AsyncValue.guard(() => build(arg));
```

This allows to let the UI decide whether it want to show the user the value is loading, or instead pick the last existing value.

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (JoachimFavre) [inline lib/viewmodels/user_centauri_points_view_model.dart:35]: After a discussion with @gruvw on a similar matter for my PR [#316](https://github.com/ProximaEPFL/proxima/issues/316), it appears that we may want to be careful with such small optimisations over Firestore's cache. It is probably better to be consistent everywhere, i.e. to trust the cache or to make optimisations everywhere.

File: test/mocks/data/post_comment.dart
Comment (JoachimFavre) [inline test/mocks/data/post_comment.dart:77]: To be consitent with the comment, we should have:
```suggestion
    ownerUsername: "username1username",
    ownerUserID: const UserIdFirestore(value: "username1ID"),
```

File: test/mocks/overrides/override_user_centauri_points_view_model.dart
Comment (JoachimFavre) [inline test/mocks/overrides/override_user_centauri_points_view_model.dart:34]: I think it is worth adding some documentation here

File: test/mocks/providers/provider_ranking.dart
Comment (JoachimFavre) [inline test/mocks/providers/provider_ranking.dart:42]: By humble curiosity, why did you wrapped this `MaterialApp` in a `ProviderScope`?

**Review by JoachimFavre — CHANGES_REQUESTED** [review]
Good job with this great PR! I have some comments, don't be afraid of their number their number is simply proportional to the size of your PR (and, they should be quick to fix!)

I just have another general comment. I feel like you no longer a `DynamicUserAvatar` and a `UserAvatar`, we can get rid of `DyanmicUserAvatar, can't we?

File: lib/utils/user_avatar_color.dart
Comment (gruvw) [reply lib/utils/user_avatar_color.dart:5]: please check previous reviews, notably https://github.com/ProximaEPFL/proxima/pull/348#discussion_r1614739378, I think it will answer your question

Comment (CHOOSEIT) [reply lib/utils/user_avatar_color.dart:5]: I think it does not make sense to leave it in the dynamic user avatar as it does not handle color anymore.

I think this is linked to this suggestion: https://github.com/ProximaEPFL/proxima/pull/348#discussion_r1614739378

File: lib/viewmodels/users_ranking_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/users_ranking_view_model.dart]: You are right !

Added in: [3347ba7](https://github.com/ProximaEPFL/proxima/pull/348/commits/3347ba7256a1066d388c2e66814259b6399a4be8)

Thank you !

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart]: I did not know 👍🏼 

I decided to remove in [243304f](https://github.com/ProximaEPFL/proxima/pull/348/commits/243304f86b9f757c4460da135fb1ba639f882e5f) the `.refreshWithCentauriPointsNumber` as the` .refresh` have a more general usage.

Thank you 💯 

Tell me what you think 👍🏼

File: test/mocks/data/post_comment.dart
Comment (CHOOSEIT) [reply test/mocks/data/post_comment.dart]: You are right! 

Fixed in :[18b20c1](https://github.com/ProximaEPFL/proxima/pull/348/commits/18b20c1ce2b4d53f54fcebdda746e1683f3668a8)

Thank you 💯

File: test/mocks/providers/provider_ranking.dart
Comment (CHOOSEIT) [reply test/mocks/providers/provider_ranking.dart]: A test was failing on a pop-up that was not finding the `ProviderScope`.

Thanks to you, I came back on this, and realised that we already had a `ProviderScope` but that was not wrapping the `MaterialApp` as we usually do, that is why I wrapped it.

Fixed in: [6b09986](https://github.com/ProximaEPFL/proxima/pull/348/commits/6b0998610759ac85c5dc49153e8ebe78e2f1ed05)

Thank you 💯

File: test/mocks/overrides/override_user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply test/mocks/overrides/override_user_centauri_points_view_model.dart:29]: You are right 💯 

Added in: [54b80e6](https://github.com/ProximaEPFL/proxima/pull/348/commits/54b80e62e6891282a6cd1a744d8a32d5e7c209a1)

Thank you !

File: lib/views/components/content/user_avatar/user_avatar.dart
Comment (CHOOSEIT) [reply lib/views/components/content/user_avatar/user_avatar.dart]: You are right.

Fixed in: [ad5cdda](https://github.com/ProximaEPFL/proxima/pull/348/commits/ad5cddab335c641e4c227995e7a579559f0aaf19)

Thank you 💯

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart:38]: You are right ! 

Added in: [5154d96](https://github.com/ProximaEPFL/proxima/pull/348/commits/5154d966387b81f654a70a210edc41deec765ae3)

Thank you !

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart]: You are completely right, however I think it relies on doing a conditional `watch` in the UI with the current state of the code, which is not desired. I think that doing a clean solution is much wider than this PR's scope and should be addressed in a separate PR, along with the potential merge of `DynamicUserAvatar` and `UserAvatar` and the removal of nullable user IDs. 👍🏼 

Tell me what you think: 💯

**Comment by CHOOSEIT** [comment]
> I just have another general comment. I feel like you no longer a `DynamicUserAvatar` and a `UserAvatar`, we can get rid of `DyanmicUserAvatar`, can't we?

You are completely right ! I answered in this comment: https://github.com/ProximaEPFL/proxima/pull/348#discussion_r1615254059

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (JoachimFavre) [reply lib/viewmodels/user_centauri_points_view_model.dart]: Mhm I see indeed. This would require a big refactor to never pass a null id; which is not a choice you took in this PR. This makes sense ok, you can stay like this.

File: lib/utils/user_avatar_color.dart
Comment (JoachimFavre) [reply lib/utils/user_avatar_color.dart:5]: Hahahah! I read Yoann's comments and it appears I have managed to miss this one and only comment. That's sad x)

Thank you, I learnt something today :D

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (JoachimFavre) [reply lib/viewmodels/user_centauri_points_view_model.dart]: This is perfect :)

(Don't worry, I did not know this morning either haha)

File: lib/views/components/content/user_avatar/user_avatar.dart
Comment (JoachimFavre) [reply lib/views/components/content/user_avatar/user_avatar.dart]: Thanks. However, why did you choose to keep the exception re-throw? I feel like it does not bring much to the code, but still increases slightly the cognitive complexity while destroying the stack trace of the original error (preventing us from knowing from where it originated).

File: test/mocks/providers/provider_ranking.dart
Comment (JoachimFavre) [reply test/mocks/providers/provider_ranking.dart]: Ah yes, I understand better now! Great, thank you 🚀

File: lib/views/components/content/user_avatar/user_avatar.dart
Comment (CHOOSEIT) [reply lib/views/components/content/user_avatar/user_avatar.dart]: Oh, I see what you mean 😄 . Moreover since we don't always have a circular value backing the the widget, is the not relevant and this is better to directly throw the error without breaking the stack trace.

Removed in: [7850437](https://github.com/ProximaEPFL/proxima/pull/348/commits/7850437df297d31ee796897e115b38325b7682e3)

Thank you 💯

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart]: I agree. 

Modified in: [0e38ef4](https://github.com/ProximaEPFL/proxima/pull/348/commits/0e38ef4bcd244327ec37406349edaacc520318b2)

Thank you ! 

Note that I made many changes, let us continue this in the main thread as I am going to ask @yoannLafore to re-review as a big chunk changed 👍🏼

**Comment by CHOOSEIT** [comment]
As I changed a big chunk since the approval of @yoannLafore, I think he should take a look at the new code.

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (CHOOSEIT) [reply lib/viewmodels/user_centauri_points_view_model.dart]: Note that I added the task of accepting a non-nullable `userId` in the dynamic user avatar view model: [#353](https://github.com/ProximaEPFL/proxima/issues/353)

**Review by yoannLafore — APPROVED** [review]
I have gone through the resolved conversations and they look good to me :+1:
I also learned a few things that I didn't know about dart.

Thanks for your great work @CHOOSEIT and @JoachimFavre :tada: 
The PR looks good to me.

File: lib/viewmodels/user_centauri_points_view_model.dart
Comment (JoachimFavre) [reply lib/viewmodels/user_centauri_points_view_model.dart]: You refactor is perfect, great job! :)

Yes, I agree with the fact that creating a new task to the backlog for the `userId` is the good choice, you've already done a lot in this PR; and this was not a choice you took in this PR. Great job :)

**Review by JoachimFavre — APPROVED** [review]
You made an amazing job at fixing my comments, I don't have anything to say on the code you modified. Great job!

I'm sorry for the refactor to move the points refresh to the dynamic profile view-model, I have to admit I did not really anticipate it would be so big. I however find your code absolutely incredible after this refactor, so thank you so much for doing it! :)

LGTM!

**Pull request merged**
